dev/ProgressBar/ProgressBar.cpp

```diff
@@ -0,0 +1,154 @@
+    m_layoutRoot.set(GetTemplateChildT<winrt::Grid>(L"LayoutRoot", controlProtected));
+    if (auto layoutRoot = m_layoutRoot.get())
+    {
+        // NOTE: You can hook up to events or property changes on your template parts here, but you have to hook up revokers also
+        // because the template can change or the control could be removed and readded to the tree, etc.
+        // (I can think of no reason you would want to handle grid loading, this is just an example of how it works.)
+        m_layoutRootLoadedRevoker = layoutRoot.Loaded(winrt::auto_revoke, { this, &ProgressBar::OnLayoutRootLoaded });
+    }
+
+    m_progressBarIndicator.set(GetTemplateChildT<winrt::Rectangle>(L"ProgressBarIndicator", controlProtected));
+    if (auto progressBarIndicator = m_progressBarIndicator.get())
+    {
+        m_progressBarIndicatorRevoker = progressBarIndicator.Loaded(winrt::auto_revoke, { this, &ProgressBar::OnLayoutRootLoaded });
+    }
+
+    UpdateStates();
+}
+
+void ProgressBar::OnLoaded(const winrt::IInspectable&, const winrt::RoutedEventArgs&)
+{
+    // TODO: things
+}
+
+void ProgressBar::OnLayoutRootLoaded(const winrt::IInspectable&, const winrt::RoutedEventArgs&)
+{
+    // TODO: things
+}
+
+void ProgressBar::OnRangeBaseValueChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    // NOTE: This hits when the Value property changes, because we called RegisterPropertyChangedCallback.
+
+    SetProgressBarIndicatorWidth();
+}
+
+void ProgressBar::OnIsIndeterminatePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    // NOTE: This hits when IsIndeterminate changes because we set MUX_PROPERTY_CHANGED_CALLBACK to true in the idl.
+
+    UpdateStates();
+}
+
+void ProgressBar::OnShowPausedPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
+}
+
+void ProgressBar::OnShowErrorPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
+}
+
+void ProgressBar::UpdateStates()
+{
+    if (ShowError())
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Error", true);
+    }
+    else if (ShowPaused() && IsIndeterminate())
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Error", true); // Paused-Indeterminate state same visual treatment as Error state
+    }
+    else if (ShowPaused())
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Paused", true);
+    }
+    else if (IsIndeterminate())
+    {
+        UpdateWidthBasedTemplateSettings();
+        winrt::VisualStateManager::GoToState(*this, L"Indeterminate", true);
+    }
+    else if (!IsIndeterminate())
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Determinate", true);
+    }
+}
+
+void ProgressBar::SetProgressBarIndicatorWidth()
+{
+    if (auto progressBarIndicator = m_progressBarIndicator.get())
+    {
+        double progressBarWidth = m_layoutRoot.get().ActualWidth();
```

Comment:
>m_layoutRoot.get() [](start = 34, length = 18)

it is good practice to check to make sure template parts exist before you use them. This could be null if they retemplate